docs/design/sparse-v2.md

```diff
@@ -0,0 +1,225 @@
+    type: SparsePatternsPathType,
+    path: RepoPathBuf,
+}
+
+pub struct SparsePatternsMapping {
+    recursive: bool,  // If false, only immediate children of src_path (files) are renamed.
+    src_path: RepoPathBuf,
+    dst_path: RepoPathBuf,
+}
+
+pub struct SparsePatterns {
+    paths: Vec<SparsePatternsPath>,
+    mappings: Vec<SparsePatternsMapping>,
+}
+
+pub trait OpStore {
+    ...
+    pub fn read_sparse_patterns(&self, id: &SparsePatternsIn) -> OpStoreResult<SparsePatterns> { ... }
+    pub fn write_sparse_patterns(&self, sparse_patterns: &SparsePatterns) -> OpStoreResult<SparsePatternsId> { .. }
+}
+```
+
+To support these more complex behaviors, a new `SparsePatterns` trait will be
+introduced, mainly to facilitate migration to the new data type at all the
+places where sparse patterns are currently used.
+
+```rust
+impl SparsePatterns {
+    pub fn to_matcher(&self) -> Box<dyn Matcher> {
+        ...
+    }
+
+    ...
+}
+```
+
+## Command Syntax
+
+Path rules can be specified on the CLI and in an editor via a `<type>:<path>`
+syntax. The editor will only support the new explicit syntax, whereas the CLI
+arguments will continue to support bare `<path>` arguments in a
+backwards-compatible way at least for a migration period.
+
+```rust
+pub fn type_string(type: SparsePatternsPathType) -> &'static str {
+    match type {
+        SparsePatternsPathType::DirInclusion => "include",
+        SparsePatternsPathType::DirExclusion => "exclude",
+        SparsePatternsPathType::Files => "files",
+        SparsePatternsPathType::File => "file",
+    }
+}
+```
+
+- `jj sparse set --add foo/bar` is equal to `jj sparse set --add include:foo/bar`
+- `jj sparse set --add exclude:foo/bar` adds a new `DirExclusion` type rule
+- `jj sparse set --exclude foo/bar` as a possible shorthand for the above
+- `jj sparse list` will print the explicit rules
+
+Paths will be stored in an ordered, canonical form which unambiguously describes
+the set of files to be included. Every `--add` command will append to the end of
+this list before the patterns are canonicalized. Whether a file is included is
+determined by the first matching rule in reverse order.
+
+For example:
+
+```txt
+include:foo
+exclude:foo/bar
+include:foo/bar/baz
+exclude:foo/bar/baz/qux
+```
+
+Produces rule set which includes "foo/file.txt", excludes "foo/bar/file.txt",
+includes "foo/bar/baz/file.txt", and excludes "foo/bar/baz/qux/file.txt". If the
+rules are subtly re-ordered, they become canonicalized to a smaller, but
+functionally equivalent form:
+
+```txt
+# Before
+include:foo
+exclude:foo/bar/baz/qux
+include:foo/bar/baz
+exclude:foo/bar
+
+# Canonicalized
+include:foo
+exclude:foo/bar
+```
+
+## Working Copy Map
+
+WARNING: This section is intentionally lacking, more research is needed.
+
+All Sparse Patterns v2 will come equipped with a default no-op mapping.
+
+```rust
+vec![WorkingCopyMap {
+    src_path: RepoPathBuf::root(),
+    dst_path: RepoPathBuf::root(),
```

Comment:
Do we want a separate type for the "client paths"?

Also, I think we need a better testing strategy to ensure that these kinds of paths are not confused. Probably making the test repo by default have a "/" -> "/client-root-for-testing/" mapping (so, just like how in Git, all tests run in a directory whose name contains a space, in order to make sure that e.g. subcommand invocations are properly quoted and escaped) would be sufficient.